src/main/java/com/limechain/trie/structure/decoded/node/DecodedNode.java

```diff
@@ -228,4 +239,218 @@ public byte[] calculateMerkleValue(Function<byte[], byte[]> hashFunction, boolea
 
         return nodeValue;
     }
+
+    /***
+     * Decodes a node value found in a proof into its components.
+     * This can decode nodes no matter their version or hash algorithm.
+     *
+     * @param encoded The byte array representing the encoded node.
+     * @return {@code DecodedNode<Nibbles, List <Byte>>} The decoded node
+     * @throws NodeDecodingException If the encoded array is null, empty, or invalid.
+     * @throws IllegalArgumentException If 'encoded' is null.
+     */
+    public static DecodedNode<Nibbles, List<Byte>> decode(byte[] encoded) {
+        ScaleCodecReader reader = new ScaleCodecReader(encoded);
+        if (encoded == null || encoded.length == 0) {
+            throw new NodeDecodingException("Invalid node value: it's empty");
+        }
+
+        int firstByte = reader.readByte() & 0xFF;
+        boolean hasChildren;
+        Boolean storageValueHashed;
```

Comment:
I'd add a NOTE for some kind of intentionality indication that `null` is a semantically valid value of this Boolean. I guess you've considered the `Optional<Boolean>` approach, although in Java that's understandably controversial and I can see why you've gone for the simpler nullable type :D